server: TestClusterConnectivity failed

[(server).TestClusterConnectivity failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2450467&tab=buildLog) on [master@e2e2cecbaec8d93a70f62951b05f5c2f30cc36de](https://github.com/cockroachdb/cockroach/commits/e2e2cecbaec8d93a70f62951b05f5c2f30cc36de):

```
      /usr/local/go/src/compress/gzip/gunzip.go:251 +0xca
  bytes.(*Buffer).ReadFrom()
      /usr/local/go/src/bytes/buffer.go:204 +0x158
  io.copyBuffer()
      /usr/local/go/src/io/io.go:395 +0x3fe
  io.Copy()
      /usr/local/go/src/io/io.go:368 +0x304
  github.com/cockroachdb/cockroach/pkg/security/securitytest.bindataRead()
      /go/src/github.com/cockroachdb/cockroach/pkg/security/securitytest/embedded.go:44 +0x2c3
  github.com/cockroachdb/cockroach/pkg/security/securitytest.test_certsClientRootKeyBytes()
      /go/src/github.com/cockroachdb/cockroach/pkg/security/securitytest/embedded.go:312 +0x92
  github.com/cockroachdb/cockroach/pkg/security/securitytest.test_certsClientRootKey()
      /go/src/github.com/cockroachdb/cockroach/pkg/security/securitytest/embedded.go:319 +0x45
  github.com/cockroachdb/cockroach/pkg/security/securitytest.Asset()
      /go/src/github.com/cockroachdb/cockroach/pkg/security/securitytest/embedded.go:455 +0x12c
  github.com/cockroachdb/cockroach/pkg/security/securitytest.RestrictedCopy()
      /go/src/github.com/cockroachdb/cockroach/pkg/security/securitytest/securitytest.go:36 +0x64
  github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.PGUrlWithOptionalClientCertsE()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/pg_url.go:95 +0x6c4
  github.com/cockroachdb/cockroach/pkg/testutils/sqlutils.PGUrlE()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/sqlutils/pg_url.go:45 +0x18f
  github.com/cockroachdb/cockroach/pkg/testutils/serverutils.OpenDBConnE()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:268 +0x1d3
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).startServer()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:421 +0x204
  github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/testserver.go:409 +0x11b
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).startServer()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:417 +0x126
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:247 +0x117

Goroutine 1165 (running) created at:
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:246 +0x21d
  github.com/cockroachdb/cockroach/pkg/server_test.TestClusterConnectivity.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/connectivity_test.go:275 +0x7d0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202

Goroutine 264 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:246 +0x21d
  github.com/cockroachdb/cockroach/pkg/server_test.TestClusterConnectivity.func2()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/connectivity_test.go:275 +0x7d0
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202
==================
    testing.go:1038: race detected during execution of test
    --- FAIL: TestClusterConnectivity/topology=5 (3.48s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestClusterConnectivity PKG=./pkg/server TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestClusterConnectivity.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -423,6 +423,8 @@ func (tc *TestCluster) startServer(idx int, serverArgs base.TestServerArgs) erro
 		return err
 	}
 
+	tc.mu.Lock()
+	defer tc.mu.Unlock()
 	tc.Conns = append(tc.Conns, dbConn)
 	return nil
 }

